File: src/ros_coverage_example/my_code.py
#!/usr/bin/env python
"""
This file implements the functionality we want in a ros-independent way
"""


class SomethingDoer:
    def __init__(self):
        self.initialised = True

    def do_thing(self, arg: str):
        if arg.startswith("test"):
            return True
        elif arg.startswith("not"):
            return False
        else:
            # untested branch
            return True

    def untested_something(self, arg: str):
        if arg.startswith("asdf"):
            return False

        return True


class OtherThingDoer:
    def __init__(self):
        self.initialised = True

    def do_other_thing(self, arg: int):
        if arg < 0:
            raise ValueError
        elif 0 <= arg <= 10:
            return True

        if arg < 10:
            # This is an untested branch
            pass
        else:
            return False

    def untested_other_thing(self, arg: int):
        if arg > 10:
            raise ValueError

        return False

File: src/my_code.cpp
#include <ros_coverage_example/my_code.hpp>
#include <stdexcept>

namespace ros_coverage_example {

bool SomethingDoer::doThing(std::string arg) const {
  if (arg.rfind("test", 0) == 0) {
    return true;
  } else if (arg.rfind("not", 0) == 0) {
    return false;
  } else {
    // untested branch
    return true;
  }
}

bool SomethingDoer::untestedSomething(std::string arg) const {
  if (arg.rfind("asdf", 0) == 0) {
    return false;
  }
  return true;
}

bool OtherThingDoer::doOtherThing(int arg) const {
  if (arg < 0) {
    throw std::invalid_argument("Argument is less than zero");
  } else if (arg >= 0 && arg <= 10) {
    return true;
  } else if (arg >= 30 && arg <= 50) {
    return true;
  } else {
    return false;
  }
}

bool OtherThingDoer::untestedOtherThing(int arg) const {
  if (arg > 10) {
    throw std::invalid_argument("Argument is greater than 10");
  }

  return false;
}

} // namespace ros_coverage_example